### PR TITLE
Features/missed throws

### LINKED ISSUE #4158: Add Pokeball miss per pokemon to pokemon_catch_worker.py
### Expected Behavior
When we throw pokeball, sometimes we miss. Usually I use up to 5 balls to catch a Zubat, then I "run" if I couldn't get. 

### Actual Behavior
With the Bot, I can catch a Zubat at 100% success rate.

### Possible solution
Change the pokemon_catch_worker.py to add a throw miss rate and maximum numbers of pokeball before "run".

### How it would help others
It would be more human like.

### PR DESCRIPTION
**Pull request by rdeeb**
## Short Description:

Adds a more human like behaviour to the **pokemon_catch_worker**, by reducing the 100% hit rate the bot had on every throw. This will add a _hit_rate_ parameter that will allow the bot to miss some shots just like a human. Also randomises a bit the waiting times so it's not always the same.

After a short period of running this changes I found the bot shows a few more pokemon spawns.

Closes #4158

### CODE REVIEW

**Comment by mjmadsen** [comment]
Would be nice if we could switch to a min/max sleep range and add them to the config. If we do generate_random_sleep_time(3), we will sleep an average of 1.8 seconds.

**Comment by DanielVolz** [comment]
I like your idea, it's really great. I tried your PR with the default 0.75 rate and this is the outcome:
screenshot 2016-08-16 23 32 06

**Comment by rdeeb** [comment]
Yeah, screwed an operator, fixing it and pushing the fix after work

**Comment by mjmadsen** [comment]
@DanielVolz You can try the inverse value for now. If you want 75% success rate, use 0.25.

**Comment by rdeeb** [comment]
Pushed the fix for the logic, now it should work with the intended value (i.e. 0.75)

**Comment by Quantra** [comment]
I like the idea =]

If you are pausing to simulate the user throwing the ball can you use the existing `action_delay()` with `catch_wait_min` and `catch_wait_max` please?

```
            # Take some time to throw the ball from config options
            action_delay(self.config.catchsim_catch_wait_min, self.config.catchsim_catch_wait_max)
```

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py [hunk 1]
```diff
@@ -110,7 +111,7 @@ def work(self, response_dict=None):
         )
 
         # simulate app
-        sleep(3)
+        action_delay(self.config.catchsim_catch_wait_min, self.config.catchsim_catch_wait_max)
```
Comment (Quantra) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: This should be app simulation not waiting to throw a ball.  The app will be quite consistent in the amount of time it takes to play an animation.  sleep(3) should be fine but if you want to expose the option it needs to be a separate option.

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py [hunk 2]
```diff
@@ -123,7 +124,7 @@ def work(self, response_dict=None):
         self._do_catch(pokemon, encounter_id, catch_rate_by_ball, is_vip=is_vip)
 
         # simulate app
-        time.sleep(5)
+        action_delay(self.config.catchsim_catch_wait_min, self.config.catchsim_catch_wait_max)
```
Comment (Quantra) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: Same here this simulates waiting for the app to play an animation not waiting for the user to throw the ball.

**Comment by Quantra** [comment]
Looking good =] please see line notes about app simulation vs humanization.

**Comment by rdeeb** [comment]
@Quantra my latest commit should solve your notes on simulation

**Comment by Quantra** [comment]
LGTM =]  Personally looking forward to this one a lot.

Something for another PR but I think we should open up app simulation delays as options to the user at some point.  Just because it's what the app does doesn't mean the bot has to or needs to so the user should be able to choose to disable them or alter them.  High risk high reward mode =]

**Comment by BreezeLo** [comment]
Wasn't this already a thing? I've been doing this for a while now, using 

`  
          "catch_throw_parameters": {
            "excellent_rate": 0.1,
            "great_rate": 0.45,
            "nice_rate": 0.27,
            "normal_rate": 0.1,
            "spin_success_rate" : 0.42
          },`

**Comment by rdeeb** [comment]
@BreezeLo what you posted will randomise the type of catch (Good, Great, Excellent, Normal, or if it has curvature or not), but the ball will hit the Pokemon always, with this change, the ball will miss the Pokemon, completely.

**Comment by BreezeLo** [comment]
@rdeeb If I understand you correctly this was already taking place with just the above settings I mentioned, which raised the fail chance from the default thus missing the pokemon and having to try again.

**Comment by rdeeb** [comment]
@BreezeLo it was not like that, there was a hard coded value that the ball always hitted the pokemon:
`hit_pokemon=1`
So this PR is to randomize that, and now the API's answer with a MISSED response, and it handles it.

**Comment by BreezeLo** [comment]
here it is

`  def generate_throw_quality_parameters(self, throw_parameters):
        throw_excellent_chance = self.catch_throw_parameters_excellent_rate
        throw_great_chance = self.catch_throw_parameters_great_rate
        throw_nice_chance = self.catch_throw_parameters_nice_rate
        throw_normal_throw_chance = self.catch_throw_parameters_normal_rate

        # Total every chance types, pick a random number in the range and check what type of throw we got
        total_chances = throw_excellent_chance + throw_great_chance \
                        + throw_nice_chance + throw_normal_throw_chance

        random_throw = random() * total_chances

        if random_throw <= throw_excellent_chance:
            throw_parameters['normalized_reticle_size'] = 1.70 + 0.25 * random()
            throw_parameters['normalized_hit_position'] = 1.0
            throw_parameters['throw_type_label'] = 'Excellent'
            return

        random_throw -= throw_excellent_chance
        if random_throw <= throw_great_chance:
            throw_parameters['normalized_reticle_size'] = 1.30 + 0.399 * random()
            throw_parameters['normalized_hit_position'] = 1.0
            throw_parameters['throw_type_label'] = 'Great'
            return

        random_throw -= throw_great_chance
        if random_throw <= throw_nice_chance:
            throw_parameters['normalized_reticle_size'] = 1.00 + 0.299 * random()
            throw_parameters['normalized_hit_position'] = 1.0
            throw_parameters['throw_type_label'] = 'Nice'
            return

        # Not a any kind of special throw, let's throw a normal one
        # Here the reticle size doesn't matter, we scored out of it
        throw_parameters['normalized_reticle_size'] = 1.25 + 0.70 * random()
        throw_parameters['normalized_hit_position'] = 0.0
        throw_parameters['throw_type_label'] = 'Normal'`

**Comment by BreezeLo** [comment]
> @BreezeLo it was not like that, there was a hard coded value that the ball always hitted the pokemon:
hit_pokemon=1
So this PR is to randomize that, and now the API's answer with a MISSED response, and it handles it.

I think you are mistaken, the variable "hit_pokemon" is only used 2 times entirely in the current build on either branch

`  C:\PokemonGo-Bot\pokemongo_bot\cell_workers\pokemon_catch_worker.py (1 hit)
	Line 411:                 hit_pokemon=1,
  C:\PokemonGo-Bot\proto\RpcSub.proto (1 hit)`

**pokemon_catch_worker.py** 
`            response_dict = self.api.catch_pokemon(
                encounter_id=encounter_id,
                pokeball=current_ball,
                normalized_reticle_size=throw_parameters['normalized_reticle_size'],
                spawn_point_id=self.spawn_point_guid,
                hit_pokemon=1,
                spin_modifier=throw_parameters['spin_modifier'],
                normalized_hit_position=throw_parameters['normalized_hit_position']
            )`

**RpcSub.proto**
`  optional bool hit_pokemon = 5;

that's it.
`

**Comment by rdeeb** [comment]
Here: https://github.com/PokemonGoF/PokemonGo-Bot/issues/4180 you will find a discussion about hit_pokemon and how changing that variable sent a new response **"4"** and it is that it **missed** the pokemon, and it was not previously handled by the bot, because the bot hit the pokemon 100% of the time.

Also in here:
```
response_dict = self.api.catch_pokemon(
                encounter_id=encounter_id,
                pokeball=current_ball,
                normalized_reticle_size=throw_parameters['normalized_reticle_size'],
                spawn_point_id=self.spawn_point_guid,
                hit_pokemon=1,
                spin_modifier=throw_parameters['spin_modifier'],
                normalized_hit_position=throw_parameters['normalized_hit_position']
            )
```
We pass a hard coded 1, that means we always hit the pokemon, in different positions of the circle yielding:

* Normal throws: Hit outside the circle in the app
* Nice throws: Hit inside the green circle when it is really big
* Great throws: Hit inside the green circle when it is at a medium size
* Excellent throws: Hit inside the green circle when it is really small

This values are calculated from this variables:
```
throw_excellent_chance = self.catch_throw_parameters_excellent_rate
throw_great_chance = self.catch_throw_parameters_great_rate
throw_nice_chance = self.catch_throw_parameters_nice_rate
throw_normal_throw_chance = self.catch_throw_parameters_normal_rate
```

But previously there was not a case when the throw will miss the pokemon, because we passed true (1) to the variable stating if we hit the Pokemon. Now that we randomly send 1 or 0 (true or false) the API answers with status 4 that is that the throw entirely missed the pokemon, i.e. you flicked your thumb to far left, and it hit grass.

**Comment by BreezeLo** [comment]
@rdeeb

> We pass a hard coded 1, that means we always hit the pokemon, in different positions of the circle yielding:

 It is impossible that the current dev branch is always hitting the Pokemon, check this image, timestamped only seconds ago from this very reply 

image

using the settings I previously quoted

**Comment by rdeeb** [comment]
@BreezeLo you are mistaken, **capture failed** means the pokemon broke free form the Pokeball. We were throwing balls at 100% hit rate.It always did a hit different XP modifiers (i.e.: Normal, Nice, Excelent, Curve balls) but the bot never missed a throw!

**Comment by BreezeLo** [comment]
@rdeeb I understand now, disregard and please, proceed :dancer:

**Comment by alexyaoyang** [comment]
@rdeeb This is a great feature! I couldn't wait for this PR to get merged so I pulled from yours directly. 

I think you missed out just a few things (or did I pull from the wrong branch?):
line 56: self.catch_throw_parameters_hit_rate = self.catch_throw_parameters.get('hit_rate', 0.8)
line 286: hit_rate = self.catch_throw_parameters_hit_rate #remove .config
line 475: action_delay(self.catchsim_catch_wait_min, self.catchsim_catch_wait_max) #remove .config

Cheers!

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py [hunk 3]
```diff
@@ -282,6 +283,7 @@ def _do_catch(self, pokemon, encounter_id, catch_rate_by_ball, is_vip=False):
         berry_id = ITEM_RAZZBERRY
         maximum_ball = ITEM_ULTRABALL if is_vip else ITEM_GREATBALL
         ideal_catch_rate_before_throw = 0.9 if is_vip else 0.35
+        hit_rate = self.config.catch_throw_parameters_hit_rate
```
Comment (RedViper9) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: Tried putting in these changes manually into current dev.

Get error:
  File "~/PokemonGo-Bot/pokemongo_bot/cell_workers/pokemon_catch_worker.py", line 319, in _do_catch
    hit_rate = self.config.catch_throw_parameters_hit_rate
AttributeError: 'dict' object has no attribute 'catch_throw_parameters_hit_rate'

**Comment by RedViper9** [comment]
@rdeeb pulled this and got various `'dict' object has no attribute (Attribute)` errors. Applying the changes  @alexyaoyang suggests results in working bot.

Thanks for this PR I think this is really necessary to humanize the bot. Even when I'm paying attention I don't always get a hit.

**Comment by satonio** [comment]
With the changes @alexyaoyang said, it works. Else it fails.

**Comment by rdeeb** [comment]
Fixed merge issues

**Comment by TheSavior** [comment]
LGTM

:+1:

**Pull request merged**